app/models/concerns/mergeable.rb

```diff
@@ -24,20 +24,28 @@ def compute_merge_status(attributes)
     end
 
     def merge(attributes)
-      new_record = new(attributes)
-      existing_record = with_discarded.find_by(id: attributes["id"])
+      retries = 0
+
+      begin
+        new_record = new(attributes)
+        existing_record = with_discarded.find_by(id: attributes["id"])
+
+        case compute_merge_status(attributes)
+        when :discarded
+          discarded_record(existing_record)
+        when :invalid
+          invalid_record(new_record)
+        when :new
+          create_new_record(attributes)
+        when :updated
+          update_existing_record(existing_record, attributes)
+        when :old
+          return_old_record(existing_record)
+        end
+      rescue ActiveRecord::RecordNotUnique
+        retries += 1
 
-      case compute_merge_status(attributes)
-      when :discarded
-        discarded_record(existing_record)
-      when :invalid
-        invalid_record(new_record)
-      when :new
-        create_new_record(attributes)
-      when :updated
-        update_existing_record(existing_record, attributes)
-      when :old
-        return_old_record(existing_record)
+        retries <= 1 ? retry : raise
```

Comment:
My natural design-for-high-throughput-brain flags a synchronous retry immediately, but I believe we won't run into any particular backpressure issues for the moment.

How can we test this?

Reply:
Well, to start, we can look at the frequency of the problem. The number of extra retries shouldn't exceed that for sure. I'm going to see if any non-production environments have the same exceptions firing organically. If so, we can deploy to one of them and observe drop-offs in exceptions and successful updates.

Ooh, we could instrument some logs or NR events to identify specific records that are affected, so we can follow up on them one-by-one in, say, Sandbox. I'll do this too.